### PR TITLE
Fix data race in nestkernel/per_thread_bool_indicator.cpp

### PR DESCRIPTION
**Pull request by jprotze**

### CODE REVIEW

File: nestkernel/per_thread_bool_indicator.h
Comment (heplesser) [inline nestkernel/per_thread_bool_indicator.h:120]:
```suggestion

  void
```

File: nestkernel/per_thread_bool_indicator.h
Comment (heplesser) [inline nestkernel/per_thread_bool_indicator.h:129]:
```suggestion

  void
```

File: nestkernel/per_thread_bool_indicator.h
Comment (heplesser) [inline nestkernel/per_thread_bool_indicator.h:132]: It is NEST convention to always write out boolean operators:
```suggestion
    if ( per_thread_status_[ tid ].is_true() and not status )
```

File: nestkernel/per_thread_bool_indicator.h
Comment (heplesser) [inline nestkernel/per_thread_bool_indicator.h:166]: We generally avoid multiple definitions in a single statement. 

```suggestion
  std::atomic< int > size_ { 0 };
  
  /** Number of per-thread indicators currently true
   *
   * are_true_ == 0 -> all are false
   * are_true_ == size_  -> all are true
   * 0 < are_true_ < size_  -> some true, some false
   */
  std::atomic< int > are_true_ { 0 };
```

Another question: Does `size_` need to be atomic, or is there any advantage to it being atomic, given that it is only set by `initialize()` and then just read later?

File: nestkernel/per_thread_bool_indicator.cpp
Comment (heplesser) [inline nestkernel/per_thread_bool_indicator.cpp:57]: We require braces also for single-statement blocks.
```suggestion
  if ( status )
  {
    are_true_ = num_threads;
  }
  else
  {
    are_true_ = 0;
  }
```

File: nestkernel/per_thread_bool_indicator.cpp
Comment (heplesser) [inline nestkernel/per_thread_bool_indicator.cpp:63]:
```suggestion
// We need two barriers here to ensure that no thread can continue and change the result
// before all threads have determined the result.
#pragma omp barrier
```

**Review by heplesser — CHANGES_REQUESTED** [review]
@jprotze Thank you for the PR. I just have a few suggestions mainly related to NEST coding conventions.

**Comment by heplesser** [comment]
@ackurth Could you benchmark this branch on a single node with microcircuit, up to 128 threads? Just to check that the change does not introduce a performance regression.

**Comment by ackurth** [comment]
> @ackurth Could you benchmark this branch on a single node with microcircuit, up to 128 threads? Just to check that the change does not introduce a performance regression.

Will do.

File: nestkernel/per_thread_bool_indicator.h
Comment (jprotze) [reply nestkernel/per_thread_bool_indicator.h:166]: I think, size_ doesn't need to be atomic

**Review by heplesser — COMMENTED** [review]
@jprotze This looks good to me now. I still wait for results from systematic benchmarks on larger models (first checks look good) and will wait for final approval until then.

**Comment by ackurth** [comment]
> @ackurth Could you benchmark this branch on a single node with microcircuit, up to 128 threads? Just to check that the change does not introduce a performance regression.

I compared the suggested changes to the current NEST Master, there is no performance regression in our usual benchmarks of the microcircuit model.

**Comment by heplesser** [comment]
> > @ackurth Could you benchmark this branch on a single node with microcircuit, up to 128 threads? Just to check that the change does not introduce a performance regression.
> 
> I compared the suggested changes to the current NEST Master, there is no performance regression in our usual benchmarks of the microcircuit model.

Excellent, then we can proceed. Thanks!

**Review by heplesser — APPROVED** [review]
With positive benchmark results, I approve.

**Review by suku248 — APPROVED** [review]
Thanks @jprotze for this PR and @ackurth for benchmarking!

**Pull request merged**
